File: tools/xdhq/Atlas/ZND/Atlas.php
<?php
/*
Copyright (C) 2018 Claude SIMON (http://q37.info/contact/).

This file is part of XDHq.

WDHq is free software: you can redistribute it and/or
modify it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

XDHq is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with XDHq. If not, see <http://www.gnu.org/licenses/>.
 */

function getXDHq() {
    if (getenv("EPEIOS_SRC") !== false) {
        switch (strtoupper(substr(php_uname('s'), 0, 3))) {
        case "WIN":
            $epeiosPath = "h:\\hg\\epeios\\";
            break;
        case "LIN":
            $epeiosPath = "/home/csimon/hg/epeios/";
            break;
        case "DAR":
            $epeiosPath = "/Users/csimon/hg/epeios/";
            break;
        default:
            die("Unknown OS !!!");
            break;
        }

        $xdhqPath = $epeiosPath . "tools/xdhq/wrappers/ZND/";
    } else {
        $xdhqPath = "phar://Atlas.phar/";
    }

    require $xdhqPath . "XDHq.php";
}

getXDHq();

set_error_handler(function ($errno, $errstr) {
    return strpos($errstr, 'Declaration of') === 0;
}, E_WARNING);

class Atlas extends Thread {
    const GUI_NONE = 0;
    const GUI_DESKTOP = 1;
    const GUI_WEB = 2;
    const GUI_DESKTOP_AND_WEB = 3;
    const GUI_DEFAULT = self::GUI_DESKTOP;

    static private function execute_(string $command) {
        if (XDHq::isWin()) {
            fclose(popen("start " . $command, 'r'));
        } else {
            fclose(popen($command . " &", 'r'));
        }
    }

    private static function launchWeb_($dir) {
        if (XDHq::isDev()) {
            self::execute_("node h:/hg/epeios/tools/xdhq/examples/common/httpd.js h:/hg/epeios/tools/xdhq/examples/common/" . $dir);
        } else {
            self::execute_("node -e \"require(require('xdhwebqnjs').fileName).launch('" . $dir . "');\"");
        }

    }

    private static function launchDesktop_($dir) {
        if (self::isDev()) {
            self::execute_("h:/hg/epeios/tools/xdhelcq/node_modules/electron/dist/electron h:/hg/epeios/tools/xdhelcq/index.js -m=h:/bin/xdhqxdh h:/hg/epeios/tools/xdhq/examples/common/" . $dir);
        } else {
            self::execute_("node -e \"require('child_process').spawnSync(require('xdhelcq').electron,[require('path').join(require('xdhelcq').path, 'index.js'),'-m=' + require('xdhqxdh').fileName, require('upath').normalize(require('path').join(require('path').resolve(__dirname),'" . $dir . "'))])\"");
        }

    }

    private static function getDefaultGUI_() {
        if (XDHq::isDev()) {
            return self::GUI_DESKTOP;
        } else {
            return self::GUI_WEB;
        }

    }

    private static function getDefaultMODE_() {
        if (XDHq::isDev()) {
            return XDHq::MODE_PROD;
        } else {
            return XDHq::MODE_DEMO;
        }

    }

    static function readAsset($path, string $dir = ".") {
        return XDHq::readAsset($path, $dir);
    }

    public function __construct($userObject) {
        $this->dom = new XDHqDOM($this);
        $this->userObject = $userObject;
    }

	public static function new(callable $callback) {
		return new Atlas(call_user_func($callback));
	}


    public static function launch(callable $callback, string $headContent = "", $gui = null, string $dir = ".") {
        global $argv;

        $mode = self::getDefaultMode_();

        if ($gui === null) {
            $gui = self::getDefaultGUI_();

            if (count($argv) > 1) {
                switch ($argv[1]) {
                case "d":
                case "desktop":
                    $gui = self::GUI_DESKTOP;
                    break;
                case "w":
                case "web":
                    $gui = self::GUI_WEB;
                    break;
                case "dw":
                case "wd":
                    $gui = self::GUI_DESKTOP_AND_WEB;
                    break;
                case "n":
                case "none":
                    $gui = self::GUI_NONE;
                    break;
                case 'W':
                    $gui = self::GUI_NONE; // The httpd server is launched externally.
                    $mode = XDHq::MODE_DEMO;
                    break;
                default:
                    die("Unknown GUI !");
                    break;
                }
            }
        }

        XDHq::launch(array("Atlas", "new"), $callback, $headContent, $mode, $dir);

        switch ($gui) {
        case self::GUI_NONE:
            break;
        case self::GUI_DESKTOP:
            self::launchDesktop_($dir);
            break;
        case self::GUI_WEB:
            self::launchWeb_($dir);
            break;
        case self::GUI_DESKTOP_AND_WEB:
            self::launchDesktop_($dir);
            self::launchWeb_($dir);
            break;
        default:
            die("Unknown gui !!!");
            break;
        }
    }

    public function run() {
        while (true) {
            $action = $this->dom->getAction($id);
            call_user_func(array($this->userObject, 'handle'), $this->dom, $action, $id);
        }
    }

    public static function createXML(string $rootTag) {
        return new XDHqXML($rootTag);
    }

    public static function createHTML(string $rootTag = "") {
        return new XDHqXML($rootTag);
    }
};
?>

File: tools/xdhq/Atlas/JRE/Atlas.java
/*
MIT License

Copyright (c) 2017 Claude SIMON (https://q37.info/s/rmnmqd49)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package info.q37.atlas;

import info.q37.xdhq.MODE;

public abstract class Atlas implements Runnable {
	public DOM dom;

	public Atlas() {
		this.dom = new DOM( info.q37.xdhq.XDH.getMode());
		new Thread( this ).start();
	}

	public abstract void handle(String action, String id );

	@Override
	public final void run() {
		info.q37.xdhq.dom.Event event = new info.q37.xdhq.dom.Event();
		try {
			for (;;) {
				dom.getAction(event);

				if ( "".equals(event.action) && isDev() )
					dom.debug_log();

				handle(event.action, event.id);
			}
		} catch (info.q37.xdhq.dom.DOM_FAAS.ThreadExitingException e) {}
		
		// System.out.println("Quitting thread!");
	}

	public static boolean isDev() {
		return info.q37.xdhq.XDH.isDev();
	}

	static public String readAsset( String path, String dir ) {
		return info.q37.xdhq.XDH.readAsset( path, dir );
	}

	public static String readAsset( String path ) {
		return readAsset( path, "" );
	}

	static public XML createXML(String rootTag) {
		return new XML(rootTag);
	}

	static public XML createHTML(String rootTag) {
		return new XML(rootTag);
	}

	static public XML createHTML() {
		return createHTML("");
	}

	private static void launchWeb(String dir) {
		try {
			ProcessBuilder processBuilder;

			if (isDev()) {
				processBuilder = new ProcessBuilder( "node", "httpd", dir );
				processBuilder.directory( new java.io.File("h:/hg/epeios/tools/xdhq/examples/common/" ) );
			} else {
				processBuilder = new ProcessBuilder( "node", "-e", "require(require('xdhwebqnjs').fileName).launch('" + dir + "');");
			}

			final Process httpd = processBuilder.start();
			Runtime.getRuntime().addShutdownHook(new Thread() {
				public void run() {
					httpd.destroy();
				}
			});
		} catch (java.io.IOException e) {
			e.printStackTrace();
			System.exit(1);
		}			
	}

	private static void launchDesktop(String dir) {
		try {
			ProcessBuilder processBuilder;

			if (isDev()) {
				processBuilder = new ProcessBuilder( "h:/hg/epeios/tools/xdhelcq/node_modules/electron/dist/electron", "index.js", "-m=h:/bin/xdhqxdh", "h:/hg/epeios/tools/xdhq/examples/common/" + dir);
				processBuilder.directory( new java.io.File("h:/hg/epeios/tools/xdhelcq") );
			} else {
				processBuilder = new ProcessBuilder( "node", "-e", "require('child_process').spawnSync(require('xdhelcq').electron,[require('path').join(require('xdhelcq').path, 'index.js'),'-m=' + require('xdhqxdh').fileName, require('upath').normalize(require('path').join(require('path').resolve(__dirname),'" + dir + "'))])");
			}

			final Process electron = processBuilder.start();
			Runtime.getRuntime().addShutdownHook(new Thread() {
				public void run() {
					electron.destroy();
				}
			});
		} catch (java.io.IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	private static final GUI getDefaultGUI() {
		if ( isDev() )
			return GUI.NONE;
		else
			return GUI.WEB;
	}

	private static final MODE getDefaultMODE() {
		if ( isDev() )
			return MODE.FAAS;
		else
			return MODE.FAAS;
	}

	private static void launch(info.q37.xdhq.XDH_SHRD.Callback callback, String headContent, String dir, GUI gui, String arg) {
		MODE mode = getDefaultMODE();

		if (gui == GUI.DEFAULT) {
			gui = getDefaultGUI();

			if (arg.length() > 0) {
				if ( arg.equals( "n" ) || arg.equals( "none" ) ) {
					gui = GUI.NONE;
				} else if ( arg.equals( "d" ) || arg.equals( "desktop" ) ) {
					gui = GUI.DESKTOP;
				} else if ( arg.equals( "W" ) ) {
					mode = MODE.FAAS;
					gui = GUI.NONE;	// The httpd server is launched externally.
				} else if ( arg.equals( "w") || arg.equals( "web" ) ) {
					gui = GUI.WEB;
				} else if ( arg.equals( "dw" ) || arg.equals( "wd") ) {
					gui = GUI.DESKTOP_AND_WEB;
				} else {
					System.err.println("Unknown gui !");
					System.exit(1);
				}
			}
		}

		info.q37.xdhq.XDH.launch(callback, headContent, mode, dir );

		switch (gui) {
		case NONE:
			break;
		case DESKTOP:
			launchDesktop(dir);
			break;
		case WEB:
			launchWeb(dir);
			break;
		case DESKTOP_AND_WEB:
			launchDesktop(dir);
			launchWeb(dir);
			break;
		default:
			System.err.println("Unknown gui !");
			System.exit(1);
			break;
		}
	}

	public static void launch(info.q37.xdhq.XDH_SHRD.Callback callback, String headContent, String dir, GUI gui, String[] args) {
		if (args.length > 0)
			launch(callback, headContent,dir, gui, args[0]);
		else
			launch(callback, headContent, dir, gui, "");
	}

	public static void launch(info.q37.xdhq.XDH_SHRD.Callback callback, String headContent, GUI gui) {
		launch(callback, headContent, ".", gui, "");
	}

	public static void launch(info.q37.xdhq.XDH_SHRD.Callback callback, String headContent) {
		launch(callback, headContent, GUI.DEFAULT);
	}

	public static void launch(info.q37.xdhq.XDH_SHRD.Callback callback) {
		launch(callback, "");
	}

	public static void broadcastAction(String action, String id ) {
		info.q37.xdhq.XDH.broadcastAction(action, id);
	}

	public static void broadcastAction(String action) {
		broadcastAction(action, "");
	}
};
